### ISSUE TITLE
SAML authentication with UAA, behind a reverse proxy

### ISSUE THREAD

**Issue by angusm43ge** [cloudfoundry/uaa#2231]
**SECURITY NOTICE: If you have found a security problem in the UAA, please do not file a public github issue. Instead, please send an email to security@cloudfoundry.org**

Thanks for taking the time to file an issue. You'll minimize back and forth and help us help you more effectively by answering all of the following questions as specifically and completely as you can.

### What version of UAA are you running?

What output do you see from `curl <YOUR_UAA>/info -H'Accept: application/json'`?
Using a downstream fork of UAA which is at version 75.14.0. /info endpoint is suppressed for security reasons.

### How are you deploying the UAA?

I am deploying the UAA

other (please explain)
packaged for kubernetes from fork, with some modifications, as part of web application.

### What did you do?
customer is running UAA with SAML auth behind a reverse proxy
TL;DR, the scenario does not work: 

### What did you expect to see? What goal are you trying to achieve with the UAA?
Looking for a view on whether SAML authentication with UAA, behind a reverse proxy, is a "supported" scenario, or whether this is more of an enhancement request

Situation here is complicated, since we are using a downstream fork of uaa, which has added its own modifications to support this.
I am raising this as a query against the upstream uaa, to get context for how to address the issue in the downstream fork.

TL;DR, the scenario does not work: the SAML login dialog is presented and completed OK, but in the ensuing callbacks, this error is seen :
```text
org.opensaml.xml.security.SecurityException: SAML message intended destination endpoint did not match recipient endpoint
```

Research elsewhere suggests that 

<https://stackoverflow.com/questions/24805895/recipient-endpoint-doesnt-match-with-saml-response>
"...at least one way to fix it is by configuring SAMLContextProviderLB instead of your current SAMLContextProviderImpl."

Referencing SpringBoot docu :
<https://docs.spring.io/spring-security-saml/docs/current/reference/pdf/spring-security-saml-reference.pdf>
page 37
"....changing
the contextProvider bean implementation in your securityContext.xml to class
org.springframework.security.saml.context.SAMLContextProviderLB:
"

"In case you use automatically generated metadata make sure to configure entityBaseURL matching
the front-end URL in your metadataGeneratorFilter bean:
"

Looking at UAA :

UAA subclasses org.springframework.security.saml.context.SAMLContextProviderImpl with org.cloudfoundry.identity.uaa.provider.saml.idp.IdpSamlContextProviderImpl, overriding

SAMLMessageContext()
populateLocalEntityId()

Looking at the recommendation in the SpringBoot docu, and following through the implications :

SpringBoot's recommended class (org.springframework.security.saml.context.SAMLContextProviderLB) also subclasses org.springframework.security.saml.context.SAMLContextProviderImpl, 

```text
adds
	private String scheme;
    private String serverName;
    private boolean includeServerPortInRequestURL;
    private int serverPort;
    private String contextPath;
	

overriding :

populateGenericContext() - calls superclass using a modified subclass of javax.servlet.http.HttpServletRequestWrapper

wrapper subclass overrides : 
getContextPath() {
getScheme()
getServerName()
getServerPort()
getRequestURI() # actually does something
isSecure() # same
```

so a UAA variant of that would need to accomodate all that as well. Kinda tricky in the details, but looks doable.

### What did you see instead?
the SAML login dialog is presented and completed OK, but in the ensuing callbacks, this error is seen :
```text
org.opensaml.xml.security.SecurityException: SAML message intended destination endpoint did not match recipient endpoint
```
Please include UAA logs if available.

**Comment by strehle** [comment]
Ok, you mentioned you have a "  fork, with some modifications, as part of web application"
So do you have adapted the Bean Definition according to 
https://docs.spring.io/spring-security-saml/docs/current/reference/html/configuration-advanced.html 
?
Because then I recommend to you, please open a PR and we check how we can use both variant with and without proxy

**Comment by angusm43ge** [comment]
I will speak with colleagues who look after the downstream fork, and get back to you on that.

 

thanks

 

From: Markus Strehle ***@***.***> 
Sent: 10 March 2023 17:01
To: cloudfoundry/uaa ***@***.***>
Cc: Macaulay, Angus J (GE Digital) ***@***.***>; Author ***@***.***>
Subject: EXT: Re: [cloudfoundry/uaa] SAML authentication with UAA, behind a reverse proxy (Issue #2231)

 


WARNING: This email originated from outside of GE. Please validate the sender's email address before clicking on links or attachments as they may not be safe.

Ok, you mentioned you have a " fork, with some modifications, as part of web application"
So do you have adapted the Bean Definition according to
https://docs.spring.io/spring-security-saml/docs/current/reference/html/configuration-advanced.html
?
Because then I recommend to you, please open a PR and we check how we can use both variant with and without proxy

**Comment by strehle** [comment]
Hi,

I have setup uaa also in docker and nginx next to and simulated this. So UAA using SAML relys on uaa.url and entityBaseURL setting, e.g.
https://github.com/cloudfoundry/uaa/blob/develop/uaa/src/test/resources/integration_test_properties.yml#L59-L61

Set both to your external name auth.saml-example.net define proxy setting and finally ensure that you have enabled / activate the reservation of host header (ProxyPassReverse ), e.g. https://www.nginx.com/resources/wiki/start/topics/examples/likeapache/
Depending on your reverse proxy there should be a setting to do this and then you will get it working - at least in my case it worked then with SAML (azure) and UAA (docker) and NGINX (docker) .

**Comment by angusm43ge** [comment]
thanks for taking time to look at this in more detail. I had also found it necessary to add proxy_set_header Host <K8SHOST>; 
I'll go back and retry this early next week.

**Comment by strehle** [comment]
close this now
